Given financial client data, let's try to predict whether a given client will go bankrupt or not.

We will use a variety of classification models to make our prediction.

**Getting started**

In [ ]:
import numpy as np
import pandas as pd
pd.set_option('max_columns', None)

import plotly.express as px

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import LinearSVC, SVC
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier

from sklearn.decomposition import PCA

import warnings
warnings.filterwarnings(action='ignore')

In [ ]:
data = pd.read_csv('/content/bankruptcy prediction.csv', header=0)

In [ ]:
data

Bankrupt?   ROA(C) before interest and depreciation before interest  \
0             1                                           0.370594          
1             1                                           0.464291          
2             1                                           0.426071          
3             1                                           0.399844          
4             1                                           0.465022          
...         ...                                                ...          
6814          0                                           0.493687          
6815          0                                           0.475162          
6816          0                                           0.472725          
6817          0                                           0.506264          
6818          0                                           0.493053          

       ROA(A) before interest and % after tax  \
0                                    0.424389   
1                                    0.538214   
2                                    0.499019   
3                                    0.451265   
4                                    0.538432   
...                                       ...   
6814                                 0.539468   
6815                                 0.538269   
6816                                 0.533744   
6817                                 0.559911   
6818                                 0.570105   

       ROA(B) before interest and depreciation after tax  \
0                                              0.405750    
1                                              0.516730    
2                                              0.472295    
3                                              0.457733    
4                                              0.522298    
...                                                 ...    
6814                                           0.543230    
6815                                           0.524172    
6816                                           0.520638    
6817                                           0.554045    
6818                                           0.549548    

       Operating Gross Margin   Realized Sales Gross Margin  \
0                    0.601457                      0.601457   
1                    0.610235                      0.610235   
2                    0.601450                      0.601364   
3                    0.583541                      0.583541   
4                    0.598783                      0.598783   
...                       ...                           ...   
6814                 0.604455                      0.604462   
6815                 0.598308                      0.598308   
6816                 0.610444                      0.610213   
6817                 0.607850                      0.607850   
6818                 0.627409                      0.627409   

       Operating Profit Rate   Pre-tax net Interest Rate  \
0                   0.998969                    0.796887   
1                   0.998946                    0.797380   
2                   0.998857                    0.796403   
3                   0.998700                    0.796967   
4                   0.998973                    0.797366   
...                      ...                         ...   
6814                0.998992                    0.797409   
6815                0.998992                    0.797414   
6816                0.998984                    0.797401   
6817                0.999074                    0.797500   
6818                0.998080                    0.801987   

       After-tax net Interest Rate  \
0                         0.808809   
1                         0.809301   
2                         0.808388   
3                         0.808966   
4                         0.809304   
...                            ...   
6814                      0

In [ ]:
data.describe()

Bankrupt?   ROA(C) before interest and depreciation before interest  \
count  6819.000000                                        6819.000000          
mean      0.032263                                           0.505180          
std       0.176710                                           0.060686          
min       0.000000                                           0.000000          
25%       0.000000                                           0.476527          
50%       0.000000                                           0.502706          
75%       0.000000                                           0.535563          
max       1.000000                                           1.000000          

        ROA(A) before interest and % after tax  \
count                              6819.000000   
mean                                  0.558625   
std                                   0.065620   
min                                   0.000000   
25%                                   0.535543   
50%                                   0.559802   
75%                                   0.589157   
max                                   1.000000   

        ROA(B) before interest and depreciation after tax  \
count                                        6819.000000    
mean                                            0.553589    
std                                             0.061595    
min                                             0.000000    
25%                                             0.527277    
50%                                             0.552278    
75%                                             0.584105    
max                                             1.000000    

        Operating Gross Margin   Realized Sales Gross Margin  \
count              6819.000000                   6819.000000   
mean                  0.607948                      0.607929   
std                   0.016934                      0.016916   
min                   0.000000                      0.000000   
25%                   0.600445                      0.600434   
50%                   0.605997                      0.605976   
75%                   0.613914                      0.613842   
max                   1.000000                      1.000000   

        Operating Profit Rate   Pre-tax net Interest Rate  \
count             6819.000000                 6819.000000   
mean                 0.998755                    0.797190   
std                  0.013010                    0.012869   
min                  0.000000                    0.000000   
25%                  0.998969                    0.797386   
50%                  0.999022                    0.797464   
75%                  0.999095                    0.797579   
max                  1.000000                    1.000000   

        After-tax net Interest Rate  \
count                   6819.000000   
mean                       0.809084   
std                        0.013601   
min                        0.000000   
25%                        0.809312   
50%                        0.809375   
75%                        0.809469   
max                        1.000000   

        Non-industry income and expenditure/revenue  \
count                                   6819.000000   
mean                                       0.303623   
std                                        0.011163   
min                                        0.000000   
25%                                        0.303466   
50%                                        0.303525   
75%                                        0.303585   
max                                        1.000000   

        Continuous interest rate (after tax)   Operating Expense Rate  \
count                            6819.000000             6.819000e+03   
mean                                0.781381             1.995347e+09   
std                                 0.012679             3.237684e+09   
min                 

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6819 entries, 0 to 6818
Data columns (total 96 columns):
 #   Column                                                    Non-Null Count  Dtype  
---  ------                                                    --------------  -----  
 0   Bankrupt?                                                 6819 non-null   int64  
 1    ROA(C) before interest and depreciation before interest  6819 non-null   float64
 2    ROA(A) before interest and % after tax                   6819 non-null   float64
 3    ROA(B) before interest and depreciation after tax        6819 non-null   float64
 4    Operating Gross Margin                                   6819 non-null   float64
 5    Realized Sales Gross Margin                              6819 non-null   float64
 6    Operating Profit Rate                                    6819 non-null   float64
 7    Pre-tax net Interest Rate                                6819 non-null   float64
 8    After-tax net Int

In [ ]:
{column: len(data[column].unique()) for column in data.columns}

{' Accounts Receivable Turnover': 1593,
 ' After-tax Net Profit Growth Rate': 6246,
 ' After-tax net Interest Rate': 3604,
 ' Allocation rate per person': 6768,
 ' Average Collection Days': 5451,
 ' Borrowing dependency': 4338,
 ' CFO to Assets': 6819,
 ' Cash Flow Per Share': 1545,
 ' Cash Flow to Equity': 6819,
 ' Cash Flow to Liability': 6819,
 ' Cash Flow to Sales': 6819,
 ' Cash Flow to Total Assets': 6819,
 ' Cash Reinvestment %': 3599,
 ' Cash Turnover Rate': 4023,
 ' Cash flow rate': 5557,
 ' Cash/Current Liability': 6816,
 ' Cash/Total Assets': 6819,
 ' Contingent liabilities/Net worth': 1855,
 ' Continuous Net Profit Growth Rate': 6270,
 ' Continuous interest rate (after tax)': 3617,
 ' Current Asset Turnover Rate': 6261,
 ' Current Assets/Total Assets': 6819,
 ' Current Liabilities/Equity': 6819,
 ' Current Liabilities/Liability': 6627,
 ' Current Liability to Assets': 6819,
 ' Current Liability to Current Assets': 6819,
 ' Current Liability to Equity': 6819,
 ' Current Liab

**Preprocessing**

In [ ]:
def preprocess_inputs(df):
  df = df.copy()

  # Drop single-value column
  df = df.drop(' Net Income Flag', axis=1)

  # Split df into X and y
  y = df['Bankrupt?']
  X = df.drop('Bankrupt?', axis=1)

  # Train-test split
  X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, shuffle=True, random_state=1)

  # Scale X
  scaler = StandardScaler()
  scaler.fit(X_train)
  X_train = pd.DataFrame(scaler.transform(X_train), index=X_train.index, columns=X_train.columns)
  X_test = pd.DataFrame(scaler.transform(X_test), index=X_test.index, columns=X_test.columns)
  return X_train, X_test, y_train, y_test

In [ ]:
X_train, X_test, y_train, y_test = preprocess_inputs(data)

In [ ]:
X_train

ROA(C) before interest and depreciation before interest  \
3392                                          -7.065972          
2755                                           0.405365          
4442                                           1.063045          
4267                                          -0.503014          
4912                                          -0.145685          
...                                                 ...          
905                                           -0.121267          
5192                                           0.353272          
3980                                           0.306876          
235                                           -2.245702          
5157                                           0.070828          

       ROA(A) before interest and % after tax  \
3392                                -7.189713   
2755                                 0.354931   
4442                                 1.174855   
4267                                -0.324003   
4912                                -0.062164   
...                                       ...   
905                                 -0.052933   
5192                                 0.448925   
3980                                 0.274365   
235                                 -2.479975   
5157                                 0.452281   

       ROA(B) before interest and depreciation after tax  \
3392                                          -7.577832    
2755                                           0.313557    
4442                                           1.177379    
4267                                          -0.454382    
4912                                          -0.098122    
...                                                 ...    
905                                           -0.160577    
5192                                           0.380411    
3980                                           0.367216    
235                                           -2.445042    
5157                                           0.066374    

       Operating Gross Margin   Realized Sales Gross Margin  \
3392                -0.353178                     -0.352216   
2755                 0.203699                      0.200630   
4442                 1.634667                      1.635325   
4267                -0.296140                     -0.283111   
4912                -0.707418                     -0.706759   
...                       ...                           ...   
905                  0.396829                      0.383409   
5192                -0.302144                     -0.300137   
3980                 0.887161                      0.889183   
235                 -0.393706                     -0.392779   
5157                 0.807107                      0.809061   

       Operating Profit Rate   Pre-tax net Interest Rate  \
3392              -27.422519                   -9.852867   
2755                0.105921                    0.077435   
4442                0.255679                    0.096169   
4267               -0.186523                    0.016814   
4912                0.011218                    0.016078   
...                      ...                         ...   
905                 0.062863                    0.032879   
5192                0.053898                    0.025100   
3980                0.157314                    0.047838   
235                -0.223142                   -0.136463   
5157                0.192417                    0.089495   

       After-tax net Interest Rate  \
3392                     -9.473508   
2755                      0.062927   
4442                      0.093795   
4267                      0.015052   
4912                      0.022094   
...                            ...   
905                       0.031040   
5192                      0.027851   
3980                      0.052564   
235                      -0.126225   
5157    

In [ ]:
y_train

3392    0
2755    0
4442    0
4267    0
4912    0
       ..
905     0
5192    0
3980    0
235     1
5157    0
Name: Bankrupt?, Length: 4773, dtype: int64

**Training original data**

In [ ]:
original_models = {
    'Logistic Regression': LogisticRegression(),
    'K-Nearest Neighbors': KNeighborsClassifier(),
    'Decision Tree': DecisionTreeClassifier(),
    'Support Vector Machine (Linear Kernel)': LinearSVC(),
    'Support Vector Machine (RBF Kernel)': SVC(),
    'Neural Network': MLPClassifier(),
    'Random Forest': RandomForestClassifier(),
    'Gradient Boosting': GradientBoostingClassifier()
}

for name, model in original_models.items():
  model.fit(X_train, y_train)
  print(name + ' trained.')

Logistic Regression trained.
K-Nearest Neighbors trained.
Decision Tree trained.
Support Vector Machine (Linear Kernel) trained.
Support Vector Machine (RBF Kernel) trained.
Neural Network trained.
Random Forest trained.
Gradient Boosting trained.


In [ ]:
original_results = []

for name, model in original_models.items():
  result = model.score(X_test, y_test)
  original_results.append(result)
  print(name + ': {:.2f}%'.format(result * 100))

Logistic Regression: 96.43%
K-Nearest Neighbors: 96.48%
Decision Tree: 94.97%
Support Vector Machine (Linear Kernel): 96.19%
Support Vector Machine (RBF Kernel): 96.58%
Neural Network: 96.09%
Random Forest: 96.43%
Gradient Boosting: 96.04%


In [ ]:
y_test.value_counts() / len(y_test)

0    0.965298
1    0.034702
Name: Bankrupt?, dtype: float64

As we can see that only one model is Support Vector Machine (RBF Kernel) that has the score (96.58%) is nearly higher than the score even if we guess all zero in y_test dataset (96.53%).

**Dimensionality Reduction (PCA)**

In [ ]:
n_components = 94

pca = PCA(n_components=n_components)
pca.fit(X_train)

X_train_reduced = pd.DataFrame(pca.transform(X_train), index=X_train.index, columns=['PC'+str(i) for i in range(1, n_components+1)])
X_test_reduced = pd.DataFrame(pca.transform(X_test), index=X_test.index, columns=['PC'+str(i) for i in range(1, n_components+1)])


In [ ]:
X_train_reduced

PC1       PC2        PC3        PC4       PC5        PC6  \
3392  21.396971 -7.427836 -10.384790  16.385044  8.410605  14.815803   
2755  -0.537250 -0.632416   1.522062   0.059505 -0.892918   0.115449   
4442  -3.777733 -0.482875   1.127331   0.334114  1.277027   0.359205   
4267   3.117700 -0.669615   1.191915  -0.644932 -1.861750   0.650279   
4912   0.849583  0.693095  -1.479401  -0.479007 -0.759656   0.081183   
...         ...       ...        ...        ...       ...        ...   
905    0.566135 -0.528486   0.117570  -0.415036  0.398719   0.994780   
5192  -2.253182  3.498919  -4.647433  -0.692899 -0.644921   1.188225   
3980  -5.630039  0.346179  -0.506075   1.037802  2.011352   0.591237   
235    7.567250 -0.172520   0.267939  -0.819849 -0.230117  -0.908622   
5157  -2.062240 -0.796774   0.953837   0.104787  0.491409   0.855032   

           PC7       PC8       PC9      PC10      PC11      PC12       PC13  \
3392 -1.072838  0.033075  2.402385 -2.696531 -4.214135  6.352764 -11.027381   
2755  0.166554 -0.332114 -0.642663 -0.422476  0.569907 -0.704901   0.165873   
4442  0.872919 -0.504780 -1.627089 -1.011487  1.137812 -0.990688   0.020866   
4267  0.850108 -0.755383 -0.427210 -0.891521  2.446697  0.372685  -0.289336   
4912 -0.799119  0.795359  1.014490  0.554192 -0.784852  0.117363   0.197798   
...        ...       ...       ...       ...       ...       ...        ...   
905  -0.124759 -0.550211 -0.998750 -0.220497  0.668082 -0.176071   0.296512   
5192 -0.218418  0.440967  0.685758  0.521140 -0.293700  0.079969  -1.141093   
3980 -1.471390 -2.752388  0.435182  0.486154 -0.318469  1.326518  -2.214204   
235   2.120545  0.029506 -0.316388  0.161774  0.262284  0.556618  -1.362690   
5157 -0.555167 -1.173928 -0.586992 -0.016522  0.919877  0.226939  -0.183864   

          PC14      PC15      PC16      PC17      PC18      PC19      PC20  \
3392  1.913793  1.606729  1.729124  0.570398 -0.789446 -0.892282 -0.320629   
2755 -0.239947  0.391633  0.419027 -0.332370 -0.234684  0.223903  0.538678   
4442  0.753910  0.551596 -0.209501  0.195654  0.323243 -0.338553  0.361258   
4267 -0.837916 -0.458419 -1.117063  1.562364 -0.621951  0.089918 -0.036900   
4912 -0.306317 -0.648215  0.198278 -0.216087 -0.563785  0.191637  0.507754   
...        ...       ...       ...       ...       ...       ...       ...   
905  -0.487469  0.525745 -0.043795 -0.090736 -0.096207  0.062817  0.579313   
5192  0.572681  0.073298  0.105754  0.095118 -1.244165  0.311395  1.247736   
3980  0.359735  1.235405  0.666536 -0.731189  0.467581  0.157602 -1.642165   
235   1.061149  0.310024 -0.363151  0.437792 -0.383468 -0.439398 -0.079670   
5157 -0.678809  1.226499  0.421790 -0.563221 -0.020129  0.066542  0.253991   

          PC21      PC22      PC23      PC24      PC25      PC26      PC27  \
3392 -1.797002 -2.299265  0.158909 -2.584296 -2.901013  1.291582  0.719369   
2755 -1.338388  0.872265  1.142810  0.173128 -0.165659  0.182387 -0.318347   
4442 -0.945430  0.968478  0.652160  0.036588  0.389052  0.179218  0.169331   
4267 -1.141743  1.467604  0.620103 -0.020184 -0.798211  1.344014 -1.028388   
4912  0.341631 -0.472753 -0.227815  0.322146  0.595560 -0.593572  0.457913   
...        ...       ...       ...       ...       ...       ...       ...   
905  -0.673308  0.638365  0.475411 -0.073659 -0.139973  0.821015 -0.406412   
5192 -0.710281 -0.465626  0.392731  0.572558  0.457306 -1.594876 -0.003512   
3980  1.390326  0.782434 -0.854762 -0.119603  0.513977 -0.101583  0.377824   
235  -1.374327  1.627859  0.399152  0.336535  0.322937  0.196119 -0.543238   
5157 -1.141191  1.891446  0.693935 -0.062868  0.545730  0.864223 -0.047809   

          PC28      PC29      PC30      PC31      PC32      PC33      PC34  \
3392  0.071927 -1.077647  1.268982  0.234510 -1.286139 -2.170726  0.026729   
2755 -0.268019 -0.151064  0.371043 -0.610176  0.540534 -0.316358  0.052687   
4442 -0.071738  0.091093 -0.075175 -0.328946  0.002957  0.206092  0.648976   
4267 -0.5459

In [ ]:
X_train_reduced.describe()

PC1           PC2           PC3           PC4           PC5  \
count  4.773000e+03  4.773000e+03  4.773000e+03  4.773000e+03  4.773000e+03   
mean  -3.154122e-17  7.885305e-17  2.093444e-18  1.586162e-16 -4.117106e-18   
std    3.608307e+00  2.666728e+00  2.301153e+00  2.240383e+00  2.170172e+00   
min   -3.173173e+01 -4.271184e+01 -2.856743e+01 -3.362950e+01 -1.580168e+01   
25%   -1.700507e+00 -1.163244e+00 -1.271392e+00 -4.503368e-01 -1.269600e+00   
50%    1.700214e-01 -1.948525e-01  9.362779e-02 -8.585801e-02 -3.095307e-01   
75%    1.753490e+00  9.014156e-01  1.325154e+00  2.939256e-01  9.136103e-01   
max    4.569358e+01  9.899954e+01  5.306129e+01  1.242471e+02  4.062732e+01   

                PC6           PC7           PC8           PC9          PC10  \
count  4.773000e+03  4.773000e+03  4.773000e+03  4.773000e+03  4.773000e+03   
mean  -3.354162e-17 -7.350314e-17  1.372369e-17  2.321397e-17  3.583859e-17   
std    1.802285e+00  1.736072e+00  1.499769e+00  1.485012e+00  1.445549e+00   
min   -2.510623e+01 -1.456150e+01 -2.596985e+01 -1.398715e+01 -2.017072e+01   
25%   -7.179408e-01 -7.511981e-01 -5.192860e-01 -6.529970e-01 -5.821846e-01   
50%   -3.674715e-02 -4.592338e-02 -2.226302e-02 -3.950109e-02 -2.732652e-02   
75%    6.784001e-01  6.907828e-01  4.891811e-01  6.050492e-01  4.992185e-01   
max    5.396654e+01  5.931159e+01  4.796535e+01  5.148651e+01  4.738088e+01   

               PC11          PC12          PC13          PC14          PC15  \
count  4.773000e+03  4.773000e+03  4.773000e+03  4.773000e+03  4.773000e+03   
mean  -5.787209e-17  1.237458e-17 -3.070384e-18 -5.303391e-18 -2.246963e-17   
std    1.394536e+00  1.338177e+00  1.329033e+00  1.276697e+00  1.223592e+00   
min   -1.742615e+01 -2.674972e+01 -1.760693e+01 -2.033150e+01 -2.768727e+01   
25%   -5.428464e-01 -5.557264e-01 -4.511933e-01 -6.623583e-01 -5.306749e-01   
50%    1.261996e-02 -1.461842e-01  1.044920e-01 -1.005780e-01  1.235171e-02   
75%    5.459625e-01  3.522767e-01  5.307571e-01  5.459004e-01  5.210342e-01   
max    4.925841e+01  4.282990e+01  2.254265e+01  2.710599e+01  2.778978e+01   

               PC16          PC17          PC18          PC19          PC20  \
count  4.773000e+03  4.773000e+03  4.773000e+03  4.773000e+03  4.773000e+03   
mean  -1.221176e-17 -1.525888e-17  5.375353e-18  1.240947e-17  5.570887e-18   
std    1.164099e+00  1.158493e+00  1.121536e+00  1.114318e+00  1.103978e+00   
min   -2.014867e+01 -1.754710e+01 -2.517376e+01 -1.798275e+01 -2.010547e+01   
25%   -3.178387e-01 -4.313930e-01 -2.703497e-01 -3.336631e-01 -3.535615e-01   
50%    3.141948e-02 -8.597171e-02 -4.222175e-02  1.833060e-03  4.125579e-03   
75%    3.651876e-01  3.530319e-01  1.561084e-01  3.283008e-01  3.993297e-01   
max    2.992306e+01  2.701543e+01  2.999181e+01  2.511310e+01  2.670018e+01   

               PC21          PC22          PC23          PC24          PC25  \
count  4.773000e+03  4.773000e+03  4.773000e+03  4.773000e+03  4.773000e+03   
mean  -6.378025e-17 -8.285385e-17 -2.795910e-17 -2.378676e-17  2.551675e-17   
std    1.094967e+00  1.070406e+00  1.042191e+00  1.022780e+00  1.016503e+00   
min   -1.265159e+01 -8.296972e+00 -7.681315e+00 -1.918518e+01 -1.363773e+01   
25%   -5.565504e-01 -5.719967e-01 -3.927890e-01 -2.630479e-01 -3.147093e-01   
50%   -6.201182e-03  1.122914e-01 -7.368688e-02 -3.821941e-02 -2.509606e-02   
75%    5.263280e-01  6.519335e-01  3.534332e-01  2.159077e-01  2.906978e-01   
max    1.713288e+01  8.447548e+00  3.212525e+01  3.757296e+01  1.919109e+01   

               PC26          PC27          PC28          PC29          PC30  \
count  4.773000e+03  4.773000e+03  4.773000e+03  4.773000e+03  4.773000e+03   
mean  -1.321196e-17 -2.466193e-17  9.554245e-18 -1.721276e-17 -1.674755e-17   
std    1.012519e+00  1.006838e+00  1.002407e+00  1.000604e+00  9.944115e-01   
min   -9.499189e+00 -9.949468e+00 -2.489048e+01 -1.445008e+01 -1.841302e+01   
25%   -4.135083e-01 -2.186903e-01 -1.356208e-01 -2.129002e-01

In [ ]:
# Show each variance of columns
pca.explained_variance_

array([1.30198766e+01, 7.11143711e+00, 5.29530377e+00, 5.01931524e+00,
       4.70964778e+00, 3.24822970e+00, 3.01394560e+00, 2.24930645e+00,
       2.20526006e+00, 2.08961284e+00, 1.94472927e+00, 1.79071697e+00,
       1.76632868e+00, 1.62995592e+00, 1.49717652e+00, 1.35512560e+00,
       1.34210617e+00, 1.25784407e+00, 1.24170572e+00, 1.21876845e+00,
       1.19895366e+00, 1.14576831e+00, 1.08616155e+00, 1.04607929e+00,
       1.03327812e+00, 1.02519478e+00, 1.01372255e+00, 1.00481891e+00,
       1.00120774e+00, 9.88854216e-01, 9.77155955e-01, 9.72997752e-01,
       9.57512828e-01, 9.55602468e-01, 9.49934087e-01, 9.11416486e-01,
       8.91201118e-01, 8.67961641e-01, 8.48680373e-01, 8.22060930e-01,
       8.03134635e-01, 7.80707837e-01, 7.32766300e-01, 6.87409209e-01,
       6.65390757e-01, 6.10480629e-01, 6.01046696e-01, 5.75296277e-01,
       5.18147916e-01, 5.01185780e-01, 4.44467008e-01, 4.29782002e-01,
       4.12656177e-01, 3.73541589e-01, 3.57877010e-01, 3.19794169e-01,
      

In [ ]:
# Variance ratio of components
pca.explained_variance_ratio_

array([1.38480306e-01, 7.56377360e-02, 5.63212164e-02, 5.33857833e-02,
       5.00921388e-02, 3.45483953e-02, 3.20565334e-02, 2.39237787e-02,
       2.34552982e-02, 2.22252664e-02, 2.06842747e-02, 1.90461893e-02,
       1.87867938e-02, 1.73363237e-02, 1.59240728e-02, 1.44132094e-02,
       1.42747339e-02, 1.33785163e-02, 1.32068677e-02, 1.29629054e-02,
       1.27521538e-02, 1.21864708e-02, 1.15524893e-02, 1.11261716e-02,
       1.09900174e-02, 1.09040424e-02, 1.07820231e-02, 1.06873233e-02,
       1.06489146e-02, 1.05175217e-02, 1.03930982e-02, 1.03488713e-02,
       1.01841725e-02, 1.01638538e-02, 1.01035645e-02, 9.69388866e-03,
       9.47887660e-03, 9.23169992e-03, 9.02662303e-03, 8.74349680e-03,
       8.54219542e-03, 8.30366244e-03, 7.79375295e-03, 7.31133180e-03,
       7.07714201e-03, 6.49311410e-03, 6.39277415e-03, 6.11889091e-03,
       5.51105699e-03, 5.33064655e-03, 4.72738177e-03, 4.57119104e-03,
       4.38903958e-03, 3.97301413e-03, 3.80640458e-03, 3.40135285e-03,
      

In [ ]:
fig = px.bar(
    x=['PC'+str(i) for i in range(1, n_components+1)],
    y=pca.explained_variance_ratio_,
    labels={'x': 'Principal Component', 'y': 'Variance Ratio'},
    color=pca.explained_variance_ratio_,
    color_continuous_scale=[(0, 'lightblue'), (1, 'darkblue')],
    title='Proportion of Variance in Principal Components'
)
fig.show()

Have a look at the graph above, we realize that the final principal components have very little information because the variance is so small and we can drop off these lower columns to reduce dimensions.

Just remain only 10 principal components has the highest variance.

In [ ]:
n_components = 10

pca = PCA(n_components=n_components)
pca.fit(X_train)

X_train_reduced = pd.DataFrame(pca.transform(X_train), index=X_train.index, columns=['PC'+str(i) for i in range(1, n_components+1)])
X_test_reduced = pd.DataFrame(pca.transform(X_test), index=X_test.index, columns=['PC'+str(i) for i in range(1, n_components+1)])


In [ ]:
fig = px.bar(
    x=['PC'+str(i) for i in range(1, n_components+1)],
    y=pca.explained_variance_ratio_,
    labels={'x': 'Principal Component', 'y': 'Variance Ratio'},
    color=pca.explained_variance_ratio_,
    color_continuous_scale=[(0, 'lightblue'), (1, 'darkblue')],
    title='Proportion of Variance in Principal Components'
)
fig.show()

**Training (Reduced Data)**

In [ ]:
reduced_models = {
    'Logistic Regression': LogisticRegression(),
    'K-Nearest Neighbors': KNeighborsClassifier(),
    'Decision Tree': DecisionTreeClassifier(),
    'Support Vector Machine (Linear Kernel)': LinearSVC(),
    'Support Vector Machine (RBF Kernel)': SVC(),
    'Neural Network': MLPClassifier(),
    'Random Forest': RandomForestClassifier(),
    'Gradient Boosting': GradientBoostingClassifier()
}

for name, model in reduced_models.items():
  model.fit(X_train_reduced, y_train)
  print(name + ' trained.')

Logistic Regression trained.
K-Nearest Neighbors trained.
Decision Tree trained.
Support Vector Machine (Linear Kernel) trained.
Support Vector Machine (RBF Kernel) trained.
Neural Network trained.
Random Forest trained.
Gradient Boosting trained.


In [ ]:
reduced_results = []

for name, model in reduced_models.items():
  result = model.score(X_test_reduced, y_test)
  reduced_results.append(result)
  print(name + ': {:.2f}%'.format(result * 100))

Logistic Regression: 96.24%
K-Nearest Neighbors: 96.43%
Decision Tree: 95.16%
Support Vector Machine (Linear Kernel): 96.43%
Support Vector Machine (RBF Kernel): 96.63%
Neural Network: 96.24%
Random Forest: 96.58%
Gradient Boosting: 96.19%


**Change in Performance After PCA**

In [ ]:
fig = px.bar(
    x=np.subtract(reduced_results, original_results),
    y=reduced_models.keys(),
    orientation='h',
    labels={'x': 'Change in Performance', 'y': 'Model'},
    color=np.subtract(reduced_results, original_results),
    color_continuous_scale=[(0, 'red'), (1, 'blue'),],
    title='Change in Model Performance After Dimensionality Reduction'
)
fig.show()